### DISCUSSION TITLE
[Feature Request] [MailHook] Convert email content into a webhook request

### DISCUSSION THREAD

**Discussion by ARHAEEM** [simple-login/app#2328] [Feature Requests]
![image](https://github.com/user-attachments/assets/40faf940-0c1a-4584-9ee8-3596ff77fd6f)

Currently, I achieve this functionality using a **Cloudflare Email Worker** with just a few lines of code. However, I believe this feature should be integrated directly into **SimpleLogin**, as it has the potential to revolutionize the email forwarding and automation space.

### Why this matters:

Imagine being able to automate email content into your preferred workflow tools such as **Zapier**, **n8n**, **Make**, **IFTTT**, and more. This opens up endless possibilities for users to manage their emails in creative and productive ways.

For example:
- Automatically organizing emails into a private **Telegram group** with categorized topics.
- Getting OTPs by emails faster than opening client emails.
- Enabling faster, easier, and more private email handling compared to IMAP in some cases.
- Utilizing services like **GitHub Actions (free tier)** to forward webhook calls directly to services like **Telegram** or custom webhooks without even a server (Server-less mail clients).

This feature could add more points to make **SimpleLogin** a unique player in the market, catering not only to privacy enthusiasts but also to automation and productivity users, or even businesses this could invite hundreds of enterprises who want to have faster and secure option to process emails these days with A.I and internal apps.

---

### Example Use Case:

- **Input:** An email is sent to your alias.
- **Output:** The email content is forwarded as a JSON webhook request to a specified endpoint.

This allows for:
- Integration with **Telegram/Discord/Viber/Slack/Line bots** to display the email in a chat.
- Automation workflows for notifications or filtering based on email content.

![Organized Emails](https://github.com/user-attachments/assets/2c644985-feac-471c-9b23-07a462eafd3f)

By implementing this, **SimpleLogin** could become the ultimate solution for secure, private, and automated email management.

### Suggested Simple Start Implementation:
1. Add a "Webhook" forwarding option for aliases.
2. Allow customization of the webhook payload to include:
   - Subject
   - Body
   - Sender
   - Attachments (optional)
   - Alias used
3. Include security measures for the webhook, such as a secret token.

Thank you for considering this feature request. I’m excited about the potential impact this could have on **SimpleLogin** and its users.

**Comment by Karek2001** [comment]
WOW!
i was thinking I'm the only one doing this with CF! xDD ,I really hope it becomes a native feature

**Comment by skirven4** [comment]
I'm looking at doing something like this. I'm wondering about pushing data from emails into Splunk/Elastic/Grafana, whatever for reporting purposes. I do have CF DNS, but I'm not sure how this would work for the Worker. Can you share the code that was written to do this from a CF worker?

**Comment by ITechGeek81** [comment]
+1

**Comment by fmilz** [comment]
Adding native webhook ingestion transforms an email alias system from a simple forwarding proxy into an event-driven integration platform. Doing this reliably in production requires careful MTA architecture.

#### Architectural Blueprint for Email-to-Webhook Ingestion
If you are planning or self-hosting this functionality, you should avoid handling HTTP dispatch synchronously within the SMTP session. Doing so introduces latency that triggers SMTP timeouts (`421` / `451`) from upstream MTAs.

Here is the standard decoupled architecture for reliable mail-to-webhook processing:

```text
[Inbound SMTP: Postfix]
        |
        v (pipe / HTTP transport)
[Edge Parser Daemon (Node / Go)] 
        |  -> Streams raw RFC 822
        |  -> Extracts MIME parts, headers, attachments
        v
[Job Queue (Redis / BullMQ)]
        |
        v
[Webhook Worker Pool]
        |  -> Evaluates routing rules (alias match, sender filter)
        |  -> Signs payload (HMAC-SHA256)
        v
[External HTTP Endpoint (n8n, Make, API)]
```

---

#### Key Technical Implementation Details

1. **MIME Stream Deserialization**:
   The edge worker must parse multipart MIME boundaries, handle quoted-printable and base64 transfer encodings, and normalize character encodings (UTF-8) across plaintext and HTML bodies. Attachments larger than standard payload limits (e.g. > 5MB) should be offloaded to temporary object storage (S3/R2/MinIO), passing signed presigned download URLs in the webhook JSON payload rather than bloating the HTTP request body.

2. **Security & Authenticity (HMAC Signing)**:
   Webhooks must include signature headers to allow receivers to verify authenticity and prevent spoofing:
   - `X-Signature-Timestamp`: Unix epoch timestamp.
   - `X-Signature-256`: `HMAC-SHA256(secret, timestamp + "." + raw_payload)`.

3. **Retry Semantics & Backoff**:
   Receiving webhooks will experience rate limits and downtime. The queue worker should implement exponential backoff (e.g. 5 retries over 24 hours) and dead-letter queue (DLQ) logging rather than dropping failed events.

---

#### Production Reference
For reference, this is how [AliasFleet](https://aliasfleet.com) architectures its webhook pipeline: incoming emails are received by Postfix, converted into structured JSON events via an async BullMQ/Redis pipeline, and dispatched to configured webhook endpoints with custom routing rules and HMAC signatures alongside standard mailbox forwarding.
